### PR TITLE
Remove old makefiles

### PR DESCRIPTION
**Pull request by yifanlu**
Proposal: either rewrite the Makefiles to be cleaner (have a global .mk that does most of the work and each individual makefile does not have tons of boilerplate) OR remove them altogether.

Why not have both? I think it confuses newbies (are the Makefiles needed? do I have to make both a cmakefile and makefile?) and also encourages copying the Makefile because either they don't know cmake or didn't see it. We end up with the same results: the old, unmaintainable Makefiles propagate.

This repo should follow best practices as it will be the first thing new developers see.

### CODE REVIEW

**Comment by yifanlu** [comment]
Done.

**Comment by yne** [comment]
Sorry, I'm not on the Vita IRC so I don't really know all the reason about all those constants rewritings but why is everyone rewriting everything in rust/cmake theses days ?

Why not simply have a `helloworld_cmake` sample, and keep/unify all the current makefiles ? (of course we will also have an `helloworld_autotools` and `helloworld_vcproj` to be fair ;-) )

I've used both solutions for ages and I'm quite sure CMake-ing theses samples is simply overkill as none of them require dependency resolving or any CMake features.

Newcomers will have to learn how the vitaSDK work, but also how to deal with the CMake `rm -rf ../build/* && cmake .. && make all send` hell.

TLDR : I don't think it's a good idea.

**Comment by yifanlu** [comment]
The problem was the old Makefile was a half-solution. It required everyone to copy huge amount of boilerplate code. It also wasn't expandable (we saw troubles when introducing safe homebrew and compression). Finally the makefile was daunting to newcomers (why do we need a velf? What's the difference between velf and self? How do I add files to vpk? Etc).

Anyways I had an issue open for a month asking for opinions and with a lack of response I went ahead and did it in cmake. Yes, it would have been possible to do everything in vanailla make too but I chose cmake because of easy dependency resolving (if app depends on libs u, v, w and lib u depends on x, and y. You can use git submodule and include_directory and you're done whereas make would have required a) porting the lib to vita, 2) writing a makefile/buildscript for vita for that lib (which would only be easy if they use autoconf or cmake), and 3) make install everything into the sdk home. And then you have to manage versioning yourself...)

Most modern libraries support cmake so porting some static lib you want is as easy as adding a git submodule and adding a cmake command. Versioning can be kept with git.

However more choices is always better! If you want to take the initiative and write a global makefile that has all the features of the cmaketoolchain (aka don't do everything as global FLAGS and don't require users to copy huge amount of boilerplate code) please let us know and I will hold off on merging this.

**Comment by xyzz** [comment]
I feel like having a comment that makefiles are DEPRECATED is bad. I also think it's better to do two copies of hello world project, one will be hello_word_make, another hello_world_cmake (I think it's ok if they are exactly the same .c files)

**Comment by yifanlu** [comment]
Sure, can you make that change or do I have to do it?

**Comment by yne** [comment]
> Makefile [...] required everyone to copy huge amount of boilerplate code

I personally see each sample as a mini-project (as if it was cloned from a git), and just like every vita project, have it own *standalone* Makefile.

> It also wasn't expandable (we saw troubles when introducing safe homebrew and compression). 

I wasn't aware of this limitation, do you have any links to this issue ?

> why do we need a velf? What's the difference between velf and self? How do I add files to vpk? Etc

Those kind of answers should be in a wiki/readme, but definitely not in a CMake/Makefile

> Anyways I had an issue open for a month asking for opinions and with a lack of response I went ahead and did it in cmake.

I didn't saw your pool, but only this issue https://github.com/vitasdk/samples/issues/17 which I was about to comment, but then, was closed/resolved the same day.

> Yes, it would have been possible to do everything in vanilla make too but I chose cmake because of easy dependency resolving

As I say earlier, most (none?) of the samples need dependencies resolving.
However, it would be personally very interested to have sample which demonstrate those CMake dependencies fetching/resolving

> If you want to take the initiative and write a global makefile that has all the features of the cmaketoolchain (aka don't do everything as global FLAGS and don't require users to copy huge amount of boilerplate code) please let us know and I will hold off on merging this.

As I said in the first quote, IMO each sample should be considered as a stand alone mini-project.
Your proposed generic Makefile belongs to the vitaSDK itself, not to the sample repository (see:https://github.com/pspdev/pspsdk/blob/master/src/base/build.mak).
I could clean all those makefiles and unifies them (minus they LIBRARY options) but I'll be unable to tests generated binaries since my Vita seems to refuse to load any *Hen since weeks... (Wich is why I've become an observer more than a contributor theses weeks)

**Comment by yifanlu** [comment]
The point is that having a big makefile for each project is not maintainable. If someone writes a Vita.mk then that would work. But in the abstance of that, we'll stick to cmake. (Of course "advanced" users can do whatever they want, this will only affect what we support in the future)

**Comment by yne** [comment]
> The point is that having a big makefile for each project is not maintainable

See: https://github.com/vitasdk/samples/pull/19/files

CmakeFiles.txt are now bigger (and so, less maintainable?) than the new Makefiles :+1: 

> It also wasn't expandable (we saw troubles when introducing safe homebrew and compression).

I still need more intel on this one. Got any samples thats fail to build/compress/relocate ?

**Comment by yifanlu** [comment]
Maintainability is not purely about code size, that was just an example. The point is this: what do we want to support as a community? Right now, we have this cmake file:

https://github.com/vitasdk/vita-toolchain/blob/master/cmake_toolchain/vita.cmake

Two commands `vita_create_self` and `vita_create_vpk` can be used together to create a VPK homebrew while `vita_create_self` and `vita_create_stubs` are used to create plugins (and export a stub lib). Look at the comments for each command to see what they can do. Specifically, it abstracts away the creation of vita artifacts into three "functions" where we can easily add arguments (let's say for example we want to generate something like encrypted homebrew. All that is required is a new argument to `vita_create_self`). Another example: for VPKs, to add files to the package, you do "-a path/to/file=path/in/vpk". Now you can argue that's bad design on our part and I would agree but it's too late to change that now. In the CMake example, the complexity of handling that is hidden away in vita.cmake, but in order to do it in a makefile, you have to either manually specify the -a arguments or do some complex parsing there (and doing so would require a Vita.mk, which again I encourage someone to create). The CMake file also tracks dependencies for there vpk add files (something an argument in your makefile won't cover).

But the big advantage, as I've discussed, is dependencies. The ability for one project to include another as a git submodule, both using cmake, and it "just works." That was the main reason I decided to create "vita.cmake" rather than "Vita.mk."

There is a reason that any reasonably complex project you find uses either Cmake or autoconf to generate makefiles rather than writing out the makefiles manually. Because as platforms and systems get more complex, people find themselves needing these sort of things. While the vita isn't at that level yet, it's good that we use what we learned from other systems.

**Comment by devnoname120** [comment]
Can I merge this or does anyone have still objections to voice?

**Comment by yne** [comment]
I still think that there is nothing wrong about makefiles for building those samples.

CMakes must be showcased in at least 1 sample (sdl,helloworld_cmake) to show how to resolve/fetch/find dependencies, yet it still useless to bloat other samples with it.

I'm totally agree @yifanlu that we need an abstraction layer (build.mk) to hide vita-* utilities API changes.

**Comment by devnoname120** [comment]
I don't agree, the only reason I see not to use CMake is if one is doing a port. This is why we keep a Makefile: people can see how it works. Also, there was [an issue](https://github.com/vitasdk/vita-toolchain/issues/79) about using CMake and nobody objected, so the decision was taken.

About the abstraction layer, it was asked several times if someone wanted to do it, but nobody wanted.
For this SDK we seem to work on a if-you-want-an-unimportant-feature-implement-it-yourself basis.

**Comment by devnoname120** [comment]
I have discussed with @yne over IRC. He isn't fond of these changes, because he thinks that CMake brings unnecessary bloat. However he accepts the merging because this is a rather small change and it isn't worth eternal discussion.
